### PR TITLE
Implement continuous queries

### PR DESCRIPTION
**Pull request by pauldix**
This is by no means everything we need for CQs, but this should be good enough to get things rolling. And it's getting too big so I think it's probably time to get eyes on it and merge.

In this PR:

- [x] Doc describing continuous query sytax and examples
- [x] Doc describing proposed architecture
- [x] Implementation of new broker to trigger periodic CQ runs
- [x] Logic for determining which CQs to run and when
- [x] Run queries and write data back out
- [x] Show continuous queries

What's still needed:

- Fanout CQs (those without a group by time)
- Tests around when CQs get run (the scheduling for recomputing previous intervals, etc)
- Checks that ensure CQs can't be created that create query->write loops
- Distributing the running of each CQ across data nodes in a cluster
- A flag in the config to disable the running of all CQs. This will be important if people accidentally create a CQ that blows up their system
- Drop continuous query

### CODE REVIEW

File: continuous_queryies.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  INTO 10m.events
+  GROUP BY time(10m)
+END
+
+-- this selects from the output of one continuous query and outputs to another series
+CREATE CONTINUOUS QUERY 1h_event_count
+ON database_name
+BEGIN
+  SELECT sum(count) as count
+  INTO 1h.events
```
Comment (dgnorton) [inline continuous_queryies.md]: @pauldix  do we need a `FROM` before the `INTO`?

File: continuous_queryies.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  GROUP BY time(10m), time(1h), *
+END
+```
+
+The `GROUP BY *` indicates that we want to group by the tagset of the points written in. The same tags will be written to the output series. The multiple aggregates in the `SELECT` clause (percentile, mean) will be written in as fields to the resulting series. The time specifies that we want multiple aggregation windows.
+
+Showing what continuous queries we have:
+
+```sql
+SHOW CONTINUOUS QUERIES
```
Comment (dgnorton) [inline continuous_queryies.md]: Would `LIST CONTINUOUS QUERIES` be more consistent?

File: continuous_queryies.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
+# Continuous Queries
+
+This document lays out continuous queries and a proposed architecture for how they'll work within an InfluxDB cluster.
+
+## Definition of Continuous Queries
+
+Continuous queries serve two purposes in InfluxDB: denormalizing data into many series or into a single series and aggregating and downsampling data. We refer to these as `fan-out`, `fan-in`, or `aggregation`. Before we get to the continuous query examples, we need to define the `INTO` syntax of queries.
```
Comment (benbjohnson) [inline continuous_queryies.md]: Can you split the sentence after the colon into a list? There's a lot of "ors" and "ands". e.g.

1. denormalizing data into many series or into a single series
2. aggregating and downsampling data

Then the next sentence lists 3 ways we refer to it but we only mentioned 2 things before.

File: continuous_queryies.md (hunk 2 above)
Comment (benbjohnson) [inline continuous_queryies.md]: :+1:

File: continuous_queryies.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+```sql
+CREATE CONTINUOUS QUERY cpu_aggregation
+ON database_name
+BEGIN
+  SELECT mean(value), percentile(80, value) as percentile_80, percentile(95, value) as percentile_95
+  INTO time().events
+  GROUP BY time(10m), time(1h), *
+END
+```
```
Comment (benbjohnson) [inline continuous_queryies.md]: Specifying multiple time groupings seems confusing to me. It seems like this should be two separate continuous queries.

**Comment by benbjohnson** [comment]
A few minor comments. Overall it looks good. One nit: remove the `y` from the filename.

File: continuous_queryies.md (hunk 4 above)
Comment (otoolep) [inline continuous_queryies.md]: Yeah, I tend to agree.

File: continuous_queryies.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+Showing what continuous queries we have:
+
+```sql
+SHOW CONTINUOUS QUERIES
+```
+
+Dropping continuous queries:
+
+```sql
+DROP CONTINUOUS QUERY <name>
```
Comment (otoolep) [inline continuous_queryies.md]: Where does the name come from? Is it auto-generated when the `INTO` command is executed? Or do you need an `AS`?

File: continuous_queryies.md (hunk 1 above)
Comment (pauldix) [inline continuous_queryies.md]: In SQL `INTO` comes before `FROM` http://www.w3schools.com/sql/sql_select_into.asp

File: continuous_queryies.md (hunk 2 above)
Comment (pauldix) [inline continuous_queryies.md]: yeah, would be. I used `SHOW` because that's what the SQL trigger syntax is.

File: continuous_queryies.md (hunk 4 above)
Comment (pauldix) [inline continuous_queryies.md]: yeah, was playing around with the idea because other people had requested it. But I think it's cleaner if they're just separate.

File: continuous_queryies.md (hunk 3 above)
Comment (pauldix) [inline continuous_queryies.md]: good call, will do

**Comment by pauldix** [comment]
@benbjohnson lol, messed up the name because I renamed it once.

File: continuous_queryies.md (hunk 5 above)
Comment (pauldix) [inline continuous_queryies.md]: The name is specified as the first part of the creation of the continuous query: `CREATE CONTINUOUS QUERY foo ON ...`

Comment (otoolep) [inline continuous_queryies.md]: Oh yeah, I was looking at the `INTO` syntax example up above and didn't see a name.

File: continuous_queryies.md (hunk 1 above)
Comment (dgnorton) [inline continuous_queryies.md]: oops...my bad.  But we do need a `FROM` after the `INTO`...right?

Comment (benbjohnson) [inline continuous_queryies.md]: Yeah, there should still be a `FROM` otherwise you don't know what measurement it's coming from.

Comment (pauldix) [inline continuous_queryies.md]: yeah, whoops! will fix :)

File: continuous_queries.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+In order to prevent cycles and endless copying of data, the following limitation is enforced on continuous queries at create time:
+
+*The output of a continuous query must go to either a different measurement or to a different retention policy.*
+
+In theory they'd still be able to create a cycle with multiple continuous queries. We should check for these and disallow.
+
+## Proposed Architecture
+
+Continuous queries should be stored in the metastore cluster wide. That is, they amount to database schema that should be stored in every server in a cluster.
```
Comment (otoolep) [inline continuous_queries.md]: Minor: I'm not sure I follow the analogy with "database schema".

File: continuous_queries.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+This document lays out continuous queries and a proposed architecture for how they'll work within an InfluxDB cluster.
+
+## Definition of Continuous Queries
+
+Continuous queries serve two purposes in InfluxDB:
+
+1. Combining many series into a single series (i.e. removing 1 or more tag dimensions to make queries more efficient)
+2. Aggregating and downsampling series
+
+The purpose of both types of continuous queries is to duplicate or downsample data automatically in the background to make querying thier results fast and efficient. Think of them as another way to create indexes on data.
```
Comment (otoolep) [inline continuous_queries.md]: Do you think "index" is a good way to think about this stuff? After all, the data created by a CQ has no direct reference back to the original data, nor makes it quicker to find the original data. Or am I missing something?

File: continuous_queries.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+### No group by time
+
+For CQs that have no `GROUP BY time` clause, they should be evaluated at the data node as part of the write. The single write should create any other writes for the CQ and submit those in the same request to the brokers to ensure that all writes succeed (both the original and the new CQ writes) or none do.
+
+I imagine the process going something like this:
+
+1. Convert the data point into its compact form `<series id><time><values>`
+2. For each CQ on the measurement and retention policy without a group by time:
+3. Run the data point through a special query engine that will output 0 or 1 data point
+4. Goto #1 for each newly generated data point
```
Comment (otoolep) [inline continuous_queries.md]: And run through steps and 2 and 3? Or just run step 1?

Comment (otoolep) [inline continuous_queries.md]: It sounds like you're trying to deal with a CQ on CQ-generated data -- a chain. Is that what you're being careful to cover here?

File: continuous_queries.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+Whenever a write came into a server, we'd look at the continuous queries and see if we needed to create new writes. If we had the two CQ examples above, we'd have to expand a single write into two more writes (one for each CQ).
+
+The first CQ would have to create a new series:
+
+```
+3 - cpu_by_region region=uswest
+```
+
+The second CQ would use the same series id as the write, but would send it to another retention policy (and thus shard).
```
Comment (otoolep) [inline continuous_queries.md]: To be clear, there is no creation of a new series ID for the second CQ?

Comment (pauldix) [inline continuous_queries.md]: That's right. Because the measurement name and tagset are the same, which are what a series ID maps to.\

File: continuous_queries.md (hunk 8 above)
Comment (pauldix) [inline continuous_queries.md]: That's right. Run through steps 1, 2, and 3. Like recursion. This makes it so CQs can be chained together.

File: continuous_queries.md (hunk 6 above)
Comment (pauldix) [inline continuous_queries.md]: Database schema is things like users, databases, measurements, and tags that exist. These are all things that are replicated via topic 0 and kept in the metastore. CQs are like this kind of schema. They're replicated cluster-wide.

File: continuous_queries.md (hunk 7 above)
Comment (pauldix) [inline continuous_queries.md]: Depends on how you think of it. Some people have said this in conversation with me. They think of them as another way to index data.

File: continuous_queries.md (hunk 6 above)
Comment (otoolep) [inline continuous_queries.md]: I guess I thought of "schema" as specifically (just) the structure of the data within the database.

Comment (pauldix) [inline continuous_queries.md]: Indexes in a database are schema as well and CQs are kind of like indexes.

File: continuous_queries.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+Before we get to the continuous query examples, we need to define the `INTO` syntax of queries.
+
+### INTO
+
+`INTO` is a method for running a query and having it output into either another measurement name, retention policy, or database. The syntax looks like this:
+
+```sql
+SELECT *
+INTO [<retention policy>.]<measurement> [ON <database>]
```
Comment (dgnorton) [inline continuous_queries.md]: @pauldix can we choose a different separator for `<retention policy>.<measurement>`?  Otherwise, the parser can't distinguish between that and a measurement name with one or more '.' characters in it.

File: continuous_queries.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+```sql
+LIST CONTINUOUS QUERIES
+```
+
+Dropping continuous queries:
+
+```sql
+DROP CONTINUOUS QUERY <name>
+ON <database>
+```
```
Comment (dgnorton) [inline continuous_queries.md]: Do we need the `ON <database>`?

File: continuous_queries.md (hunk 10 above)
Comment (pauldix) [inline continuous_queries.md]: This should be the same query language elsewhere. If they have a `.` in the column name or measurement name, they need to wrap it in double quotes.

File: continuous_queries.md (hunk 11 above)
Comment (pauldix) [inline continuous_queries.md]: they are associated with a specific database on creation

**Comment by pauldix** [comment]
Ok, this thing is ready for some eyes. Have a look at the PR info at the top and let me know what you think. @benbjohnson, @dgnorton

File: database.go [hunk 12]
```diff
@@ -19,7 +19,8 @@ import (
 type database struct {
 	name string
 
-	policies map[string]*RetentionPolicy // retention policies by name
+	policies          map[string]*RetentionPolicy // retention policies by name
+	continuousQueries []*ContinuousQuery          // continuous queries by name
```
Comment (dgnorton) [inline database.go]: Comment says by name but it's just an array of pointers.  Sorted by name?  Or, should this be a `map[string]*ContinuousQuery`?

File: database.go [hunk 13]
```diff
@@ -994,6 +1005,35 @@ func (db *database) MeasurementAndSeries(name string, tags map[string]string) (*
 	return idx, idx.seriesByTags(tags)
 }
 
+// SereiesByID returns the Series that has the given id.
+func (d *database) SeriesByID(id uint32) *Series {
+	return d.series[id]
+}
+
+// Names returns all measurement names in sorted order.
+func (d *database) Names() []string {
```
Comment (dgnorton) [inline database.go]: I think this should be called `MeasurementNames`.  However, I wrote `func (db *database) MeasurementNames()`already and stupidly iterated through `db.measurements` keys.  Delete this func and fix mine?

File: cmd/influxd/run.go [hunk 14]
```diff
@@ -70,6 +71,9 @@ func Run(config *Config, join, version string, logWriter *os.File) *influxdb.Ser
 		}
 		log.Printf("data node #%d listening on %s", s.ID(), config.DataAddr())
 
+		// broker needs the server to occasionally run continuous queries
+		b.RunContinuousQueryLoop(s)
```
Comment (otoolep) [inline cmd/influxd/run.go]: No guarantee `b` is non-nil here.

File: cmd/influxd/config.go [hunk 15]
```diff
@@ -109,6 +109,32 @@ type Config struct {
 	Logging struct {
 		File string `toml:"file"`
 	} `toml:"logging"`
+
+	ContinuousQuery struct {
+		// when continuous queries are run we'll automatically recompute previous intervals
+		// in case lagged data came in. Set to zero if you never have lagged data. We do
+		// it this way because invalidating previously computed intervals would be insanely hard
+		// and expensive.
+		RecomputePreviousN int `toml:"recompute_previous_n"`
```
Comment (otoolep) [inline cmd/influxd/config.go]: The style in the rest of the config file (as far as I can see) is that `-` is used as the word delimiter for keys within a config section.

File: httpd/handler.go [hunk 16]
```diff
@@ -85,6 +85,10 @@ func NewHandler(s *influxdb.Server, requireAuthentication bool, version string)
 			"ping",
 			"GET", "/ping", h.servePing,
 		},
+		route{ // Tell data node to run CQs that should be run
+			"process_continuous_queries",
+			"POST", "/process_continuous_queries", h.serveProcessContinuousQueries,
```
Comment (otoolep) [inline httpd/handler.go]: You may need to add the gzip (false?) flag here, due to some recent changes by @corylanou .

File: server.go [hunk 17]
```diff
@@ -2757,9 +2810,272 @@ func HashPassword(password string) ([]byte, error) {
...
+		return tx.saveDatabase(db)
+	})
+
+	return nil
+}
+
+// RunContinuousQueries will run any continuous queries that are due to run and write the
+// results back into the database
+func (s *Server) RunContinuousQueries() error {
+	// s.mu.RLock()
```
Comment (otoolep) [inline server.go]: Do you mean to leave this uncommented?

File: influxql/ast.go [hunk 18]
```diff
@@ -680,6 +683,76 @@ func (s *SelectStatement) walkForTime(node Node) bool {
 	}
 }
 
+// GroupByIterval extacts the time interval, if specified.
+func (s *SelectStatement) GroupByInterval() (time.Duration, error) {
+	// return if we've already puled it out
```
Comment (toddboom) [inline influxql/ast.go]: `puled` -> `pulled`

**Comment by toddboom** [comment]
overall, looks pretty solid to me.

File: broker.go [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+		TriggerFailurePause: 1 * time.Second,
+	}
+	b.Broker = messaging.NewBroker()
+	return b
+}
+
+func (b *Broker) RunContinuousQueryLoop(d dataNodeStore) {
+	b.server = d
+	b.done = make(chan struct{})
+	go b.continuousQueryLoop()
```
Comment (benbjohnson) [inline broker.go]: Pass in `b.done` to `continuousQueryLoop()` as an argument since you close and unset it in `Close()`. You can get in a situation where you're selecting from a `nil` `b.done`.

File: broker.go [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+		// if no error, we're all good
+		err := b.requestContinuousQueryProcessing()
+		if err == nil {
+			return
+		}
+		log.Printf("broker cq: error hitting data node: %s: %s\n", b.currentCQProcessingNode.URL, err.Error())
+
+		// reset and let the loop try the next data node in the cluster
+		b.currentCQProcessingNode = nil
+		<-time.After(100 * time.Millisecond)
```
Comment (benbjohnson) [inline broker.go]: Move magic number to constant.

File: broker.go [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+	}
+}
+
+func (b *Broker) requestContinuousQueryProcessing() error {
+	// Send request.
+	cqUrl := copyURL(b.currentCQProcessingNode.URL)
```
Comment (benbjohnson) [inline broker.go]: `cqUrl` -> `cqURL`

File: cmd/influxd/config.go [hunk 22]
```diff
@@ -109,6 +109,32 @@ type Config struct {
...
+		// will be a minute past the previous 10m bucket of time before lagged data is picked up
+		ComputeRunsPerInterval int `toml:"compute_runs_per_interval"`
+		// ComputeNoMoreThan paired with the RunsPerInterval will determine the ceiling of how many times smaller
+		// group by times will be computed. For example, if you have RunsPerInterval set to 10 and this setting
+		// to 1m. Then for a group by time(1m) will actually only get computed once per interval (and once per PreviousN).
+		// If you have a group by time(5m) then you'll get five computes per interval. Any group by time window larger
+		// than 10m will get computed 10 times for each interval.
+		ComputeNoMoreThan Duration `toml:"compute_no_more_than"`
+	} `toml:"continuous_queries"`
 }
```
Comment (benbjohnson) [inline cmd/influxd/config.go]: Can you add blank lines between the fields and the next field's comment?

**Comment by benbjohnson** [comment]
Overall looks good. Just a couple comments. :ship: after fixes.

File: broker.go [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+		// Check if broker is currently leader.
+		if b.Broker.IsLeader() {
+			b.runContinuousQueries()
+		}
+
+		// Wait for a timeout or until done.
+		select {
+		case <-b.done:
+			return
+		case <-time.After(1 * time.Second):
```
Comment (otoolep) [inline broker.go]: Can we lift this to a constant too?

File: database.go [hunk 24]
```diff
@@ -994,6 +1005,35 @@ func (db *database) MeasurementAndSeries(name string, tags map[string]string) (*
 	return idx, idx.seriesByTags(tags)
 }
 
+// SereiesByID returns the Series that has the given id.
```
Comment (otoolep) [inline database.go]: Nit: typo in comment.

File: influxql/ast.go [hunk 25]
```diff
@@ -680,6 +683,76 @@ func (s *SelectStatement) walkForTime(node Node) bool {
 	}
 }
 
+// GroupByIterval extacts the time interval, if specified.
```
Comment (otoolep) [inline influxql/ast.go]: extacts -> extracts

**Comment by otoolep** [comment]
I don't see any work related to "No group by time" queries, which the design doc proposed would be done at the time of writing. Is that being dropped, or to do?

**Comment by pauldix** [comment]
@otoolep that's a todo. See the first item on the list of "What's still needed" at the top of this PR.

**Comment by otoolep** [comment]
Ah yes, I see that.

**Comment by pauldix** [comment]
Updated based on feedback. Let me know if it all looks good @otoolep and @benbjohnson

File: influxql/engine.go [hunk 26]
```diff
@@ -314,7 +314,7 @@ loop:
 	for _, row := range rows {
 		for _, values := range row.Values {
 			t := time.Unix(0, values[0].(int64))
-			values[0] = t.UTC().Format(time.RFC3339Nano)
+			values[0] = t.UTC()
```
Comment (otoolep) [inline influxql/engine.go]: Is this change deliberate?

File: cmd/influxd/run.go [hunk 27]
```diff
@@ -90,6 +90,9 @@ func Run(config *Config, join, version string, logWriter *os.File) (*messaging.B
 			go a.ListenAndServe()
 		}
 
+		// broker needs the server to occasionally run continuous queries
+		b.RunContinuousQueryLoop(s)
```
Comment (otoolep) [inline cmd/influxd/run.go]: How can you be sure `b` will be non-nil? If you look at line 46 it's checking if b is non-nil. There will be no broker object on a data-only node, right?

File: influxql/engine.go (hunk 26 above)
Comment (pauldix) [inline influxql/engine.go]: Yes it's deliberate. The rows shouldn't have string times yet. They should get turned into strings when serialized.

File: cmd/influxd/run.go (hunk 27 above)
Comment (pauldix) [inline cmd/influxd/run.go]: Fixed in 72f30b0bccd7b8584bb3e0d7c45c20649a185f1c

File: influxql/engine.go (hunk 26 above)
Comment (corylanou) [inline influxql/engine.go]: I was the one that made that change but now that I understand the behavior of the system better, clearly was the wrong spot.  We'll have to add custom marshaling to the influxql.Row struct to get time to marshal correctly.  Are we going to allow storing of time in actual data values as well?  If so, then we will have to have knowledge of all of that from the meta store when we marshal anyway.

Comment (pauldix) [inline influxql/engine.go]: We should allow storing time as a field value, but it's not a priority for now. Right now it's just int, float64, bool, string, and bytes

File: cmd/influxd/run.go (hunk 27 above)
Comment (otoolep) [inline cmd/influxd/run.go]: Still doesn't look right. If the node is only Broker, the continuous query loop will never run. And since the continuous query loop only runs on the Raft leader, it means it would never run if the cluster uses dedicated Raft nodes.

**Comment by pauldix** [comment]
Ok, now it's ready for final review @otoolep, @benbjohnson

**Comment by otoolep** [comment]
OK, having the data nodes provide their URLs during replica-registration makes sense to me -- it should work well, as far as I can tell. @benbjohnson is a better guy to understand if there are any other system issues involved here, such as change-of-URLs causing problems, which could be problematic

But it does seem like an effective way for the Raft leader to be able to reach out to a random data node when it needs CQs run.

So +1 on a green build.
